feat: Initial feature-engineering-on-fabric single-tech sample check-in (#652)
Azure-Samples/modern-data-warehouse-dataops: single_tech_samples/fabric/feature_engineering_on_fabric/README.md

# Introduction

This article will guide you to build a feature engineering system based on Azure Machine Learning Managed Feature Store and Microsoft Fabric.

![Architecture](./images/fs_architecture.png)
As the diagram shows above, running on Microsoft Fabric, a data pipeline lands, ingests and transforms the incoming data. The transformed data is registered as features in Azure Machine Learning Managed Feature Store. The features are then can be used for model training and inferencing. In the meantime, the data lineage of both the data pipeline and features are tracked and monitored by Azure Purview.

> ## Getting Started

Author: Done! :- )

![workspace_02](./images/data_pipeline/workspace_02.jpg)

##### Step 2: Create a Lakehouse

Then within this created workspace, create a new Lakehouse named as '[prefix]_lh' by clicking the 'New' button on the top of the page. Please note that [prefix] is the prefix you defined in the prerequisites section.

##### Step 3: Import the sample notebooks

After the Lakehouse is created, please click the 'New' button again, and select 'Import notebook' from the menu. Navigate to `src/notebooks` folder under the repo, and select all the notebooks to import:

![import all notebooks](./images/import_all_notebooks.png)

##### Step 4: Add the Created Lakehouse to the notebooks

After the notebooks are imported, please open the notebook named as 'data_cleansing'，'data_ingestion'，'data_transformation','data_validation'，‘EDA','feature_set_registration' respectively, and add the Lakehouse created in the previous step to the notebook. Please note that the Lakehouse name should be '[prefix]_lh', where [prefix] is the prefix you defined in the prerequisites section.

![add lakehouse](./images/add_lh_to_notebook.png)

> #### Create an Azure Resource Group

Reviewer: Rather than asking user to deploy all the resources manually, I think the best is to just add a bicep deployment file for all the Azure resources, and list all the resources that are being deployed.

Author: Hi Anuj, 
Thanks for the suggestion! 
We had some discussion around the IAC topic last week, and now @siliang-j-1225 is helping to add resources deployment logics in relevant sample notebooks, which follows the pattern shared in Azure ML managed feature store samples, the goal is to let users complete the whole e2e process in notebook environment alone without switching between deployment scripts and notebooks.

Create a new resource group named as [prefix]rg, where [prefix] is the prefix you defined in the prerequisites section.

> #### Azure App Service

Reviewer: I think this would require some additional details. I am in the app service and don't have any idea what configuration to choose. Also, please add reference links to the docuemtation.

Author: Thanks Anuj! Bo has confirmed that the App Service is not required anymore, we'll remove this in later commits, thanks! :- )

Reviewer 2: Thanks for point out this. i will remove this section and app service code.

Reviewer 2: updated!

Create a new Azure App Service named as [prefix]as, where [prefix] is the prefix you defined in the prerequisites section.

#### Azure Data Lake Storage Gen2

Create an Azure Data Lake Storage Gen2 account named as [prefix]sa, where [prefix] is the prefix you defined in the prerequisites section. For the detailed information, please refer to: [Create a storage account to use with Azure Data Lake Storage Gen2](https://learn.microsoft.com/en-us/azure/storage/blobs/create-data-lake-storage-account)

#### Azure Purview

Create an Azure Purview account named as [prefix]pv, where [prefix] is the prefix you defined in the prerequisites section.

#### Azure Machine Learning Managed Feature store

Create an Azure Machine Learning Managed Feature store by running the following Azure CLI:

> ```az ml feature-store create --subscription <subscription id> --resource-group <resource group> --location <location> --name <location>```

Reviewer: Is there a specific reason why we are creating this using CLI? For this, we need to mention that the "ml" extension needs to be added first.
```
az extension add --name ml
```

> ```az ml feature-store create --subscription <subscription id> --resource-group <resource group> --location <location> --name <location>```

Reviewer: ```suggestion
```az ml feature-store create --subscription <subscription id> --resource-group <resource group> --location <location> --name <feature store name>```
```

  
  ![data_pipeline_06](./images/data_pipeline/data_pipeline_06.png)

- Step 5, Return to the Source tab of the Copy data Activity within the ForEach Activity. In the connection, select the HTTP link that we just created. The Connection type should be HTTP. Then, enter '@item()' in the Relative URL, and select Binary as the File format. This completes the configuration of the Source in the Copy data Activity.
  
  ![data_pipeline_07](./images/data_pipeline/data_pipeline_07.png)

- Step 6, Switch to the Destination tab and make the following configurations as shown in the image. Please note, if you have not created a Lakehouse, you need to create a new Lakehouse named as '[prefix]_lh' in the workspace first, then select this Lakehouse here. If you have already created a Lakehouse, then you can directly select the Lakehouse you have created here. The file path here is the second parameter 'landing_path' that we just created, so our Copy data Activity will download the data to the corresponding folder in our Lakehouse.

>   - File Path: __@pipeline().parameters.landing_path__

Reviewer: The File Path instructs user to use "landing_path" but the image shows "SavePath" which is confusing.

Reviewer 2: I changed this before. It seems that I forgot to update this picture. Thank you for pointing out that. and just has been updated.

  ```text
  Name: client_secret
  Type: String
  Value: `@pipeline().parameters.client_secret`
  ```

- Associate actual notebooks with the activities

  Within the 'Settings' tab, select our current workspace and then choose the right Notebook file.

  ```text
  Data Ingestion       -> data_ingestion
  Data Cleansing       -> data_cleansing
  Data Transformation  -> data_transformation
  Feature Registration -> feature_set_registration
  ```
  
  ![data_pipeline_15](./images/data_pipeline/data_pipeline_15.png)

## Execute and verify

Once the setup is complete, you can trigger the pipeline and verify the results.

### Trigger the data pipeline

Access the Fabric workspace and locate the item categorized as a `Data pipeline`.

![data pipeline](./images/pipeline_image.png)

In order to trigger the data pipeline, choose this data pipeline, and click `Run`.

![data pipeline](./images/data_pipeline.png)

In the pane, you can check the input and output, which gives the more detailed execution result.

![pipeline run](./images/pipeline_run.png)

### Verify the data lineage in Purview

TBD

### Verify the features in Feature Store

> TBD

Reviewer 3: ```suggestion
If the pipeline executes successfully, you can verify the features registered in the Azure ML Managed Feature Store.
<img width="949" alt="image" src="https://github.com/Azure-Samples/modern-data-warehouse-dataops/assets/44962887/97cf6291-65e6-400a-92d8-50263777933f">
```

Reviewer: Thanks Siliag, can we also add some additional description of what it means.

#### Azure Machine Learning Managed Feature store

Create an Azure Machine Learning Managed Feature store by running the following Azure CLI:

```az ml feature-store create --subscription <subscription id> --resource-group <resource group> --location <location> --name <feature store name>```

For other ways to create a feature store, please refer to the online document:
[Create a Minimal Feature Store](https://learn.microsoft.com/en-us/azure/machine-learning/tutorial-get-started-with-feature-store?view=azureml-api-2&tabs=SDK-track#create-a-minimal-feature-store).

### Data Source Preparation

In this demo the data from the [New York City Taxi & Limousine Commission](https://www.nyc.gov/site/tlc/about/tlc-trip-record-data.page) Yellow Taxi Trip Records, which is public. We used Records for 2022. This includes anonymized travel details like departure points, destinations, times, distances, and costs. The data, in conjunction with Taxi Zone Maps and Lookup Tables, aids in various research fields. It helps identify frequent pickup and drop-off zones in the city.

We provide two options to get the sample data:

> - Option 1: ADLS (Azure Data Lake Storage) or Blob storage.

Reviewer: For simplicity, I would recommend to just use Option 2 as default. We can mention option 1 in the footnote as an alternate approach without providing all the details.

Reviewer 2: remove option 1 from the doc makes it more simpler and keep it as a footnote info. Thank you for your advice

```yaml
runtime_version: '1.1'
spark_conf:
  - spark.fsd.client_id: <sp-client-id>
  - spark.fsd.tenant_id: <sp-tenant-id>
  - spark.fsd.subscription_id: <subscription-id>
  - spark.fsd.rg_name: <feature-store-resouce-group>
  - spark.fsd.name: <feature-store-name>
  - spark.fsd.fabric.tenant: <fabric-tenant-name> # Fetch from Fabric base URL, like https://<fabric-tenant-name>.powerbi.com/
  - spark.fsd.fabric.workspace: <fabric-workspace>
  - spark.fsd.fabric.lakehouse: <fabric-lakehouse>
  - spark.fsd.purview.account: <purview-account-name>

```

After this is done, click __Save__ to save the environment, and then __Publish__. This may take a few minutes to finish.

To apply the environment, you can set the newly created environment as 'default' in the Fabric workspace settings page.

![pip](./images/fabric_env_2.png)

Or you can apply to each specific notebook in the notebook edition page.

![pip](./images/fabric_env_1.png)

> ### Model Training and Inferencing Setup

Reviewer: This seems a bit out-of-place. Can we not add it later, when we start talking about the Feature registration.

In addition to the main flow, there are optional steps for performing 'exploratory data analysis' and 'data validations' (illustrated by dotted lines in the diagram). These features are currently not covered as part of the step-by-step guide, but the notebooks are available in the repo for reference.

## Environment setup

### Prerequisites

- Access to a [Microsoft Fabric enabled tenant](https://learn.microsoft.com/fabric/admin/fabric-switch#enable-for-your-tenant). You can sign up for Fabric (Preview) trials for a period of 60 days by following [this link](https://learn.microsoft.com/fabric/get-started/fabric-trial).
- Access to an Azure subscription. You can sign up for a free trial by following [this link](https://azure.microsoft.com/free/).
- Permissions to [create a service principal](https://learn.microsoft.com/entra/identity-platform/howto-create-service-principal-portal#register-an-application-with-microsoft-entra-id-and-create-a-service-principal) in Microsoft Entra ID.

> ### Microsoft/Azure resources

Reviewer 4: The "/" seems a bit odd. Better just "Microsoft and Azure resources". Or simply Required Resources?

Reviewer: Changed to "Required resources"

### Microsoft Fabric setup

1. Create a Microsoft Fabric workspace

   After signing into Microsoft Fabric, the very first step is to create a workspace. Workspaces contain the Fabric items such as lakehouses, warehouses, and reports. For detailed instructions about creating a Fabric workspace, refer to [this documentation](https://learn.microsoft.com/fabric/data-warehouse/tutorial-create-workspace).

   Please note that you need to select 'trial' under the 'license mode' in the advanced options, as currently only this type supports all the Microsoft Fabric features required for this sample.

2. Create a Fabric lakehouse

   Within the workspace, create a new lakehouse by clicking the 'New' button on the top of the page. Refer to [this documentation](https://learn.microsoft.com/fabric/onelake/create-lakehouse-onelake#create-a-lakehouse) for detailed instructions.

3. Import the sample notebooks

>    After the lakehouse is created, go back to the workspace. Click the 'New' button again and select 'Import notebook' from the menu. Navigate to `src/notebooks` folder under the repo, and select all the notebooks to import:

Reviewer 3: ```suggestion
   After the lakehouse is created, go back to the workspace. Click the 'New' button from the **Data Engineering or the Data Science** homepage and select 'Import notebook' from the menu. Navigate to `src/notebooks` folder under the repo, and select all the notebooks to import:
```
added detail here as in PowerBI homepage the 'Import notebook' option does not show up.

Reviewer: Updated

   ![pip](./images/fabric_env_2.png)

   Or you can apply to each specific notebook in the notebook edition page.

   ![pip](./images/fabric_env_1.png)

## Data pipeline setup

This is the main data pipeline which contains all the activities required for data landing, ingestion, cleansing, transformation, and feature registration. Start by creating a new data pipeline in the workspace.

![data_pipeline_01](./images/data_pipeline/data_pipeline_01.png)

### Data landing activity

As mentioned in the introduction, a 'ForEach' activity is used for data landing. This activity downloads multiple files from a public blob storage. Add the 'ForEach' activity to the data pipeline [^1].

![data_pipeline_02](./images/data_pipeline/data_pipeline_02.png)

The configuration of this 'ForEach' activity follows a series of steps as below.

- Add copy data activity

  Select the 'ForEach' activity that was just created and click '+' to add a 'Copy data' activity.

  ![data_pipeline_03](./images/data_pipeline/data_pipeline_03.png)

- Define source and destination for 'Copy data' activity

>   Select the 'Copy data' activity and switch to the 'Source' tab. Choose 'external' as the 'Data store type'. For 'Connection', click '+' to create a new HTTP link. On the creation page, enter <https://stmdwpublic.blob.core.windows.net/> as the server URL. Click the 'Create' button to complete the setup.

Reviewer 4: Where is this currently hosted? https://stmdwpublic.blob.core.windows.net/

Reviewer: It's hosted in our external DTD subscription :-) I have put a lock on the resource group.

- Configure the 'Notebook' activities

  Rename each notebook activity by selecting it, clicking 'General' tab and updating the name of the 'Notebook' activity. Name them as 'Data Ingestion', 'Data Cleansing', 'Data Transformation', and 'Feature Registration' respectively.

  For each notebook activity, add the following 'Base parameters' in the Settings' tab:

  ```text
  Name: client_secret
  Type: String
  Value: `@pipeline().parameters.client_secret`
  ```

- Associate actual notebooks with the activities

  Within the 'Settings' tab, select our current workspace and then choose the right Notebook file.

  ```text
  Data Ingestion       -> data_ingestion
  Data Cleansing       -> data_cleansing
  Data Transformation  -> data_transformation
  Feature Registration -> feature_set_registration
  ```
  
  ![data_pipeline_15](./images/data_pipeline/data_pipeline_15.png)

## Execute and verify

Once the setup is complete, you can trigger the pipeline and verify the results.

### Trigger the data pipeline

Access the Fabric workspace and locate the item categorized as a `Data pipeline`.

![data pipeline](./images/pipeline_image.png)

> In order to trigger the data pipeline, choose this data pipeline, and click `Run`.

Reviewer 4: Its might be obvious for folks who have worked with Service Principals before given the name "client_secret", but this is missing a step to populate the client_secret pipeline parameter with the service principal password/secret

Reviewer: Updated
